### PR TITLE
Adding cred scan step in PR validation and Ci build

### PR DESCRIPTION
**Pull request by sima-zhu**
# All SDK Contribution checklist:
The PR is to add credscan into pipelines:
1. PR validation: https://dev.azure.com/azure-sdk/public/_build/results?buildId=1327591&view=logs&jobId=bc67675d-56bf-581f-e0a2-208848ba68ca
2. Ci build(Storage): [to be continued](https://dev.azure.com/azure-sdk/public/_build/results?buildId=1327596&view=logs&j=bc67675d-56bf-581f-e0a2-208848ba68ca)
This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

File: eng/pipelines/templates/jobs/ci.yml [hunk 1]
```diff
@@ -68,6 +68,13 @@ jobs:
         displayName: 'Use Python 3.6'
         inputs:
           versionSpec: '3.6'
+      - task: securedevelopmentteam.vss-secure-development-tools.build-task-credscan.CredScan@2
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:71]: Do we also need to publish the results? https://github.com/Azure/azure-sdk-for-net/blob/main/eng/pipelines/aggregate-reports.yml#L171

File: eng/pipelines/templates/jobs/ci.yml [hunk 2]
```diff
@@ -68,6 +68,28 @@ jobs:
         displayName: 'Use Python 3.6'
         inputs:
           versionSpec: '3.6'
+      - pwsh: |
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:71]: When this isn't a PR we will want this to scan the entire service directory.

File: eng/pipelines/templates/jobs/ci.yml [hunk 3]
```diff
@@ -68,6 +68,28 @@ jobs:
...
+          New-Item -Path "$(Build.sourcesdirectory)" -Name "changedFiles" -ItemType "directory"
+          git diff $targetBranch HEAD --name-only | ForEach-Object {Copy-Item -Path $_ -Destination "$(Build.sourcesdirectory)/changedFiles"}
+      - task: securedevelopmentteam.vss-secure-development-tools.build-task-credscan.CredScan@2
+        displayName: Scanning for credentials
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:75]: We should move this to the analyze stage.

File: eng/pipelines/templates/jobs/ci.yml [hunk 4]
```diff
@@ -68,6 +68,28 @@ jobs:
         displayName: 'Use Python 3.6'
         inputs:
           versionSpec: '3.6'
+      - pwsh: |
+          New-Item -Path "$(Build.sourcesdirectory)" -Name "changedFiles" -ItemType "directory"
+          git diff $targetBranch HEAD --name-only | ForEach-Object {Copy-Item -Path $_ -Destination "$(Build.sourcesdirectory)/changedFiles"}
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:73]: FYI- I think we are at a stage where we should have a common helper to find the list of changed files as this approach works sometimes it doesn't always.

Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:73]: Where is targetBranch defined?

File: eng/pipelines/templates/jobs/ci.yml [hunk 5]
```diff
@@ -190,6 +190,40 @@ jobs:
           failOnStderr: false
       - template: /eng/common/pipelines/templates/steps/eng-common-workflow-enforcer.yml
 
+  - job: "ComplianceTools"
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:193]:
```suggestion
  - job: Compliance
```

File: eng/pipelines/templates/jobs/ci.yml [hunk 6]
```diff
@@ -190,6 +190,40 @@ jobs:
...
+    pool:
+      name: azsdk-pool-mms-win-2019-general
+      vmImage: MMS2019
+    steps:      
+      - pwsh: |
+          New-Item -Path "$(Build.sourcesdirectory)" -Name "changedFiles" -ItemType "directory"
+          Write-Host "$(System.PullRequest.TargetBranch)"
+          git diff "$(System.PullRequest.TargetBranch)" HEAD --name-only | ForEach-Object {Copy-Item -Path $_ -Destination "$(Build.sourcesdirectory)/changedFiles"}
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:201]: I don't think we should copy the files but instead just find the directory they are in and add that directory to the list.

File: eng/pipelines/templates/jobs/ci.yml [hunk 7]
```diff
@@ -190,6 +190,40 @@ jobs:
...
+          Get-ChildItem "$(Build.sourcesdirectory)/changedFiles"
+        displayName: Put changed files to folder
+        condition: eq(variables['Build.Reason'],'PullRequest')
+      - task: securedevelopmentteam.vss-secure-development-tools.build-task-credscan.CredScan@2
+        displayName: Scanning for credentials
+        inputs:
+          toolMajorVersion: V2
+          toolVersion: latest
+          ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:210]: I wouldn't do this conditioning here I would keep it all in the pre-step above and have this only read a variable that contains the list of directories to scan.

File: eng/pipelines/templates/jobs/ci.yml (hunk 6 above)
Comment (sima-zhu) [reply eng/pipelines/templates/jobs/ci.yml]: Then we will scan the entire folder even if we only touch some of the files. It is not the fastest way we can do.

Comment (weshaggard) [reply eng/pipelines/templates/jobs/ci.yml]: Copying the files will cause all sorts of other potential issues so I'd like to avoid that if we can.

File: eng/pipelines/templates/jobs/ci.yml (hunk 4 above)
Comment (sima-zhu) [reply eng/pipelines/templates/jobs/ci.yml]: https://github.com/Azure/azure-sdk-tools/issues/2613

File: eng/pipelines/templates/jobs/ci.yml [hunk 8]
```diff
@@ -191,6 +190,44 @@ jobs:
...
+    pool:
+      name: azsdk-pool-mms-win-2019-general
+      vmImage: MMS2019
+    steps:      
+      - pwsh: |
+          New-Item -Path "$(Build.sourcesdirectory)" -Name "credscan.tsv"
+          if ("$(Build.Reason)" -eq 'PullRequest') {
+            New-Item -Path "$(Build.sourcesdirectory)" -Name "changedFiles" -ItemType "directory"
+            Write-Host "origin/$(System.PullRequest.TargetBranch)"
+            (git diff "origin/$(System.PullRequest.TargetBranch)" HEAD --name-only) | ForEach-Object { Copy-Item $_ -Destination "$(Build.sourcesdirectory)\changedFiles"}
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:203]: I would still like to see wat this looks like if we don't copy the files but instead just scan all the directories?

File: eng/pipelines/templates/jobs/ci.yml [hunk 9]
```diff
@@ -191,6 +190,55 @@ jobs:
...
+    pool:
+      name: azsdk-pool-mms-win-2019-general
+      vmImage: MMS2019
+    steps:      
+      - pwsh: |
+          New-Item -Path "$(Build.sourcesdirectory)" -Name "credscan.tsv"
+          if ("$(Build.Reason)" -eq 'PullRequest') {
+            New-Item -Path "$(Build.sourcesdirectory)" -Name "changedFiles" -ItemType "directory"
+            (git diff "origin/$(System.PullRequest.TargetBranch)" HEAD --name-only | Split-Path)
+            | ForEach-Object { Set-Content -Path "$(Build.SourcesDirectory)\credscan.tsv" -Value $_}
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:203]: This looks like it will only write the last directory to the file because it keeps overwritting the file.

File: eng/pipelines/templates/jobs/ci.yml (hunk 8 above)
Comment (sima-zhu) [reply eng/pipelines/templates/jobs/ci.yml]: 1. PR validation (**5 mins** run): https://dev.azure.com/azure-sdk/public/_build/results?buildId=1327591&view=logs&jobId=bc67675d-56bf-581f-e0a2-208848ba68ca
2. Ci build(Storage) (**15 mins**): [to be continued](https://dev.azure.com/azure-sdk/public/_build/results?buildId=1327596&view=logs&j=bc67675d-56bf-581f-e0a2-208848ba68ca)

Comment (sima-zhu) [reply eng/pipelines/templates/jobs/ci.yml]: @weshaggard 

Storage takes 4 mins longer than build job. What do you think about it?

File: eng/pipelines/templates/jobs/ci.yml [hunk 10]
```diff
@@ -109,7 +109,6 @@ jobs:
 
       - task: AzureArtifacts.manifest-generator-task.manifest-generator-task.ManifestGeneratorTask@0
         displayName: 'Upload Package SBOM'
-        condition: succeededOrFailed()
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:112]: Looks like a incorrect merge. You should add back this condition.

File: eng/ignore-links.txt [hunk 11]
```diff
@@ -4,4 +4,4 @@ https://github.com/Azure/azure-digital-twins/blob/private-preview/Documentation/
 https://github.com/Azure/azure-sdk-for-net-pr/tree/feature/IoT-ADT/sdk/digitaltwins/Azure.DigitalTwins.Core/samples/DigitalTwinsClientSample/DTDL/DigitalTwins
 https://github.com/Azure/azure-sdk-for-net-pr/blob/feature/IoT-ADT/sdk/iot/Azure.Iot.DigitalTwins/samples/DigitalTwinServiceClientSample/DTDL/Relationships/HospitalRelationships.json
 https://www.nuget.org/packages/Azure.ResourceManager.DigitalTwins
-https://contoso.azureedge.net/urlsigning/test?expires=2145916800&amp;keyid:key1&amp;signature=iTsrLX9rVAIJkSahBA_j5o9Azf5-j331ohxDR1Gx2js=
+https://contoso.azureedge.net/urlsigning/test?expires=2145916800&amp;keyid:key1&amp;signature=iTsrLX9rVAIJkSahBA_j5o9Azf5-j331ohxDR1Gx2js=
```
Comment (weshaggard) [inline eng/ignore-links.txt:7]: undo the changes to this file.

File: eng/pipelines/templates/jobs/ci.yml [hunk 12]
```diff
@@ -112,7 +112,6 @@ jobs:
         condition: succeededOrFailed()
         inputs:
           BuildDropPath: $(Build.ArtifactStagingDirectory)
-
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:115]: Lets add this blank line back.

File: eng/pipelines/templates/jobs/ci.yml [hunk 13]
```diff
@@ -191,6 +190,14 @@ jobs:
...
 
+  - job: Compliance
+    pool:
+      name: azsdk-pool-mms-win-2019-general
+      vmImage: MMS2019
+    steps:      
+      - template: /eng/common/pipelines/templates/steps/credscan.yml
+        parameters:
+          BaselineFilePath: $(Build.sourcesdirectory)/eng/dotnet.gdnbaselines
   - template: /eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
```
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:201]: Please add a blank line to help with readability.

**Review by weshaggard — APPROVED** [review]
Lets please make sure that the pipelines are passing before we merge this.
